packaging/build_rpm.py

```diff
@@ -0,0 +1,375 @@
+########
+# Copyright (c) 2017 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+import argparse
+import logging
+import sys
+import shutil
+import urllib2
+from datetime import datetime
+from os import chdir, listdir
+from os.path import (
+        abspath,
+        basename,
+        dirname,
+        getmtime,
+        join as path_join,
+        split as path_split,
+        )
+from subprocess import check_call, check_output, CalledProcessError
+
+
+MOCK = '/usr/bin/mock'
+LOCAL_REPO_PATH = '~/mock_repo'
+DEPENDENCIES_FILE = '{spec_file}.dependencies'
+RESULT_DIR = '/var/lib/mock/epel-7-x86_64/result'
+
+SCRIPT_DIR, SCRIPT_NAME = path_split(abspath(__file__))
+SSH_CONFIG_FILE = path_join(SCRIPT_DIR, 'ssh_config.tmp')
+
+RPMS_DIR = SCRIPT_DIR
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(SCRIPT_NAME)
+
+
+def main(args):
+    parser = argparse.ArgumentParser(
+            description="Build RPM using mock",
+            )
+    parser.add_argument(
+            'spec_file', type=argparse.FileType('r'),
+            help="which RPM to build",
+            )
+
+    args = parser.parse_args(args)
+
+    spec_file = abspath(args.spec_file.name)
+    packaging_dir, spec_file_name = path_split(spec_file)
+    # We assume that the spec always lives one dir level below the source
+    source = dirname(packaging_dir)
+
+    # Fetch any defined Sources (this does not include the repo containing the
+    # spec file itself, the local copy is to be used.
+    chdir(source)
+    sources = get_sources(spec_file)
+    sources += get_dependency_urls(spec_file)
+    logger.info(sources)
+    for url in sources:
+        download_if_newer(url)
+
+    chdir(SCRIPT_DIR)
+
+    if has_cmd('mock'):
+        # run locally
+        rpm = build(source, spec_file_name)
+        shutil.copy(rpm, SCRIPT_DIR)
+    elif has_cmd('vagrant'):
```

Comment:
I'd put all of this in a `build_in_vagrant` function (which I would also split into 3-5 parts) - it's a pretty large and self contained piece of logic

Reply:
made the `build_in_vagrant()` function but I'm not convinced splitting that out further will make it more readable.
It comfortably fits on a page (but OK, my font size is small...), and all of it that's not already factored out seems fairly self-explanatory to me.
LMK if this doesn't convince you and I'll have a go